File: 0701-0800/0707-design-linked-list.js
function Node(val) {
  this.val = val;
  this.next = null;
}
/**
 * Initialize your data structure here.
 */
var MyLinkedList = function() {
  this.head = new Node(0);
};

/**
 * Get the value of the index-th node in the linked list. If the index is invalid, return -1.
 * @param {number} index
 * @return {number}
 */
MyLinkedList.prototype.get = function(index) {
  let i = 0;
  let cur = this.head.next;
  while (cur) {
    if (i++ === index) return cur.val;
    cur = cur.next;
  }

  return -1;
};

/**
 * Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list.
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtHead = function(val) {
  const node = new Node(val);
  node.next = this.head.next;
  this.head.next = node;
};

/**
 * Append a node of value val to the last element of the linked list.
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtTail = function(val) {
  let pre = this.head;
  while (pre.next) {
    pre = pre.next;
  }
  pre.next = new Node(val);
};

/**
 * Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted.
 * @param {number} index
 * @param {number} val
 * @return {void}
 */
MyLinkedList.prototype.addAtIndex = function(index, val) {
  let i = 0;
  let pre = this.head;
  while (pre) {
    if (i++ === index) {
      let node = new Node(val);
      node.next = pre.next;
      pre.next = node;
      break;
    }
    pre = pre.next;
  }
};

/**
 * Delete the index-th node in the linked list, if the index is valid.
 * @param {number} index
 * @return {void}
 */
MyLinkedList.prototype.deleteAtIndex = function(index) {
  let i = 0;
  let pre = this.head;
  while (pre.next) {
    if (i++ === index) {
      pre.next = pre.next.next;
      break;
    }
    pre = pre.next;
  }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(val)
 * obj.addAtTail(val)
 * obj.addAtIndex(index,val)
 * obj.deleteAtIndex(index)
 */


// var obj = new MyLinkedList();
// console.info(obj);
// obj.addAtHead(1);
// console.info(obj);
// obj.addAtTail(3);
// console.info(obj);
// obj.addAtIndex(1, 2);
// console.info(obj);
// console.info(obj.get(1));
// obj.deleteAtIndex(1);
// console.info(obj.get(1));

var obj = new MyLinkedList();
console.info(obj);
obj.addAtTail(1);
console.info(obj);
console.info(obj.get(0));

File: 0701-0800/0707-design-linked-list.cc

#include <iostream>
using namespace std;

class MyLinkedList {
    struct ListNode {
        int val;
        ListNode *next;
        ListNode() : val(0), next(nullptr) {}
        ListNode(int val): val(val), next(nullptr) {}
    };
    ListNode *head;

public:
    MyLinkedList() : head(new ListNode()) {}
    
    int get(int index) {
        int i = 0;
        ListNode *cur = head->next;
        while (cur) {
            if (i++ == index) {
                return cur->val;
            }
            cur = cur->next;
        }
        return -1;
    }
    
    void addAtHead(int val) {
        ListNode *node = new ListNode(val);
        node->next = head->next;
        head->next = node;
    }
    
    void addAtTail(int val) {
        ListNode* pre = head;
        while (pre->next) {
            pre = pre->next;
        }
        pre->next = new ListNode(val);
    }
    
    void addAtIndex(int index, int val) {
        int i = 0;
        ListNode *pre = head;
        while (pre) {
            if (i++ == index) {
                ListNode *node = new ListNode(val);
                node->next = pre->next;
                pre->next = node;
                return;
            }
            pre = pre->next;
        }
    }
    
    void deleteAtIndex(int index) {
        int i = 0;
        ListNode *pre = head;
        while (pre->next) {
            if (i++ == index) {
                ListNode *t = pre->next;
                pre->next = pre->next->next;
                delete t;
                return;
            }
            pre = pre->next;
        }
    }
};

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList* obj = new MyLinkedList();
 * int param_1 = obj->get(index);
 * obj->addAtHead(val);
 * obj->addAtTail(val);
 * obj->addAtIndex(index,val);
 * obj->deleteAtIndex(index);
*/

int main() {
    MyLinkedList *obj = new MyLinkedList();
    obj->addAtHead(1);
    obj->addAtTail(3);
    obj->addAtIndex(1,2);
    cout << obj->get(1) << endl;
    obj->deleteAtIndex(1);
    cout << obj->get(1) << endl;
    return 0;
}